### PR TITLE
Unit tests on Android

### LINKED ISSUE #208: Run Android unit tests on a device
Similar to Realm Cores setup: https://github.com/realm/realm-core/tree/master/test/android

### PR DESCRIPTION
**Pull request by kneth**
Closes #208 

ABI (x86) and API (16) are hard-coded, and it requires a rooted device or an emulator.

@emanuelez @alebsack @cmelchior @beeender

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -46,7 +47,17 @@ if(REALM_ENABLE_SYNC)
     target_link_libraries(tests realm-sync realm-sync-server)
 endif()
 
-target_link_libraries(tests realm-object-store ${PLATFORM_LIBRARIES})
+if(REALM_PLATFORM STREQUAL "Android")
+  target_link_libraries(tests
+    realm-object-store
+    ${PLATFORM_LIBRARIES}
+    android
+    log
```
Comment (bdash) [inline tests/CMakeLists.txt]: Add these to `PLATFORM_LIBRARIES` in CompilerFlags.cmake instead?

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -46,7 +47,17 @@ if(REALM_ENABLE_SYNC)
     target_link_libraries(tests realm-sync realm-sync-server)
 endif()
 
-target_link_libraries(tests realm-object-store ${PLATFORM_LIBRARIES})
+if(REALM_PLATFORM STREQUAL "Android")
```
Comment (emanuelez) [inline tests/CMakeLists.txt]: You can simply use `if(ANDROID)`

File: tests/android_main.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <jni.h>
+
+#include <cerrno>
+#include <iostream>
+#include <fstream>
+#include <string>
+
+#include <android/log.h>
+#include <android/native_activity.h>
```
Comment (emanuelez) [inline tests/android_main.cpp]: Not needed. We're not creating an app, but just an Android elf file

Comment (emanuelez) [inline tests/android_main.cpp]: Actually... I don't think this whole file is needed at all.

Comment (kneth) [inline tests/android_main.cpp]: Right, it was added by mistake.

**Comment by zaki50** [comment]
Unit test crashed (Segmentation fault) on Android API 10 x86 emulator (worked fine on API 23 emulator).
Is this a known limitation?

I think unit test should work on all emulators whose API level is 9 or newer.

**Comment by zaki50** [comment]
ah, I didn't read the description of this PR...

**Comment by kneth** [comment]
@zaki50 I have a bit lazy :-( I need to enable PIE and API 9 doesn't support that. So next API level was 16 in R10e.

File: README.md [hunk 4]
```diff
@@ -91,6 +91,16 @@ If you have installed Clang as an external Xcode toolchain (using the `install-x
...
 ```
 
+### Android
+
+It requires a root device or an emulator:
+
+```
+make
+adb push tests/tests /data
+adb shell /data/tests
```
Comment (kneth) [inline README.md]: I only tried with an emulator running API 22. The goal of running the unit tests is to ensure that we don't introduce stuff which isn't compatible with Android.

File: tests/util/test_file.cpp [hunk 5]
```diff
@@ -51,10 +51,14 @@ TestFile::TestFile()
         const char* dir = getenv("TMPDIR");
         if (dir && *dir)
             return dir;
+#ifdef REALM_ANDROID
+        return "/data/local/tmp";
```
Comment (beeender) [inline tests/util/test_file.cpp]: Is this necessary? On my emulator, the `TMPDIR` is set to `/data/local/tmp`

Comment (kneth) [inline tests/util/test_file.cpp]: It is not set on my emulator. Please consider it as extra safety :-)

**Comment by zaki50** [comment]
@kneth Many tests on linux is failing with following log.

Is this PR affecting tests on linux?

```
/home/jenkins/workspace/m_realm-object-store_PR-254-45TWLDZZSMAKN2SPZFYHLLQ6SSJ5YCOQUHCS2PHUBKMOI634FZLQ/tests/handover.cpp:56: FAILED:
	due to unexpected exception with message:
	  Unable to open a realm at path '/data/local/tmp/realm.0teWyy.management':
	  make_dir() failed: No such file or directory.	
```

**Comment by beeender** [comment]
@zaki50 It seems you have `REALM_ANDROID` defined somewhere?

**Comment by zaki50** [comment]
@beeender Not on my machine, on CI https://ci.realm.io/blue/organizations/jenkins/realm%2Frealm-object-store/detail/PR-254/2/pipeline/29

File: tests/util/test_file.cpp [hunk 6]
```diff
@@ -51,10 +51,14 @@ TestFile::TestFile()
         const char* dir = getenv("TMPDIR");
         if (dir && *dir)
             return dir;
+#ifdef REALM_ANDROID
```
Comment (bdash) [inline tests/util/test_file.cpp]: `REALM_ANDROID` is always defined. It should be checked using `#if REALM_ANDROID` rather than `#ifdef REALM_ANDROID`.

**Comment by kneth** [comment]
@beeender @zaki50 My mistake. @bdash pointed out what the issue is.

**Pull request merged**